source/source.go

```diff
@@ -16,72 +16,161 @@ package source
+	// Docs: https://developer.box.com/reference/get-folders-id-items/
+	if c.BatchSize == nil || *c.BatchSize < 1 || *c.BatchSize > 1000 {
+		return ErrInvalidBatchSize
+	}
+
 	return nil
 }
 
 func NewSource() sdk.Source {
-	// Create Source and wrap it in the default middleware.
-	return sdk.SourceWithMiddleware(&Source{})
+	return sdk.SourceWithMiddleware(&Source{
+		config: Config{
+			DefaultSourceMiddleware: sdk.DefaultSourceMiddleware{
+				// disable schema extraction by default, as the source produces raw payload data
+				SourceWithSchemaExtraction: sdk.SourceWithSchemaExtraction{
+					PayloadEnabled: lang.Ptr(false),
+				},
+			},
+		},
+	})
 }
 
 func (s *Source) Config() sdk.SourceConfig {
 	return &s.config
 }
 
-func (s *Source) Open(_ context.Context, _ opencdc.Position) error {
-	// Open is called after Configure to signal the plugin it can prepare to
-	// start producing records. If needed, the plugin should open connections in
-	// this function. The position parameter will contain the position of the
-	// last record that was successfully processed, Source should therefore
-	// start producing records after this position. The context passed to Open
-	// will be cancelled once the plugin receives a stop signal from Conduit.
+func (s *Source) Open(ctx context.Context, position opencdc.Position) error {
+	sdk.Logger(ctx).Info().Msg("Opening Box source")
+
+	var err error
+	s.position, err = ParseSDKPosition(position)
+	if err != nil {
+		return fmt.Errorf("error parsing sdk position: %w", err)
+	}
+
+	if s.client == nil {
+		s.client, err = box.NewHTTPClient(s.config.Token)
+		if err != nil {
+			return fmt.Errorf("error creating http client for box: %w", err)
+		}
+	}
+
+	// Verify folder exists
+	if s.config.ParentID != 0 {
+		isFolder, err := s.client.VerifyFolder(ctx, s.config.ParentID)
+		if err != nil || !isFolder {
+			return fmt.Errorf("error verifying folder id: %w", err)
+		}
+	}
+
+	s.ch = make(chan opencdc.Record, 5)
+	s.workersWg = &sync.WaitGroup{}
+
+	// Start worker
+	s.workersWg.Add(1)
+	go func() {
+		NewWorker(
+			s.client,
+			s.config,
+			s.position,
+			s.ch,
+			s.workersWg,
+		).Start(ctx)
+	}()
+
 	return nil
 }
 
-func (s *Source) ReadN(context.Context, int) ([]opencdc.Record, error) {
-	// ReadN is the same as Read, but returns a batch of records. The connector
-	// is expected to return at most n records. If there are fewer records
-	// available, it should return all of them. If there are no records available
-	// it should block until there are records available or the context is
-	// cancelled. If the context is cancelled while ReadN is running, it should
-	// return the context error.
-	return []opencdc.Record{}, nil
+func (s *Source) ReadN(ctx context.Context, n int) ([]opencdc.Record, error) {
+	if s.ch == nil {
+		return nil, ErrSourceClosed
+	}
+
+	records := make([]opencdc.Record, 0, n)
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case r, ok := <-s.ch:
+		if !ok {
```

Comment:
It doesn't look like the channel above is ever closed. We should make sure we close it when the worker is done and propagate the error from the worker to this method (`ReadN`).